src/vision/status_quo/alan_hates_writing_a_stream.md

```diff
@@ -0,0 +1,194 @@
+
+### Implementing `Stream`
+
+Implementing a `Stream` means writing async code in a way that doesn't _feel_ like the `async fn` that Alan has written so far. He needs to write a `poll` function and it has a lot of unfamiliar concepts:
+
+- `Pin`
+- State machines
+- `Wakers`
+
+Unsure of what the final code will look like, he starts with:
+
+```rust
+struct SigningFile;
+
+impl Stream for SigningFile {
+    type Item = Result<Vec<u8>, Error>;
+    
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context)
+        -> Poll<Self::Item>
+    {
+ 
+    }
+}
+```
+
+#### Pin :scream:
+
+First, he notices `Pin`. Alan wonders, "Why does `self` have bounds? I've only ever seen `self`, `&self`, and `&mut self` before". Curious, he reads the [`std::pin`](https://doc.rust-lang.org/std/pin/struct.Pin.html) page, and a bunch of jargon about pinning data in memory. He also reads that this is useful to guarantee that an object cannot move, and he wonders why he cares about that. The only example on the page explains how to write a ["self-referential struct"][self-ref], but notices it needs `unsafe` code, and that triggers an internal alarm in Alan: "I thought Rust was safe..."
+
+Eventually (how?), Alan realizes that the types he's depending on are `Unpin`, and so he doesn't need to worry about the unsafe stuff. It's just a more-annoying pointer type.
+
+[self-ref]: https://doc.rust-lang.org/std/pin/index.html#example-self-referential-struct
+
+#### State Machine
+
+With `Pin` hopefully ignored, Alan next notices that in the imperative style he wanted originally, he didn't need to explicitly keep track of state. The state was simply the imperative order of the function. But in a `poll` function, the state isn't saved by the compiler. Alan finds blog posts about the dark ages of Futures 0.1, when even `Future` was written with a "state machine".
+
+He thinks about his stream's states, and settles on the following structure:
+
+```rust
+struct SigningFile {
+    state: State,
+    file: ReaderStream,
+    sig: Signature,
+}
+
+enum State {
+    File,
+    Sign,
+}
+```
+
+Now he tries to write the `poll_next` method, checking readiness of individual steps and proceeding to the next state, while grumbling away the weird `Pin` noise:
+
+```rust
+todo!("is this worth showing in the story, or just noise?")
+```
+
+It turns out it was more complicated than Alan thought (the author made this same mistake). The `digest` method of `Signature` is `async`, _and_ it consumes the signature, so the state machine needs to be adjusted. The signature needs to be able to be moved out, and it needs to be able to store a future from an `async fn`. Trying to figure out how to represent that in the type system was difficult (expand?), but Alan eventually learns to just store a `Pin<Box<dyn Future>>`, wondering if the `Pin` there is important.
```

Comment:
At the risk of the story getting too long, I do think it's worth expanding how Alan was able to land on `Pin<Box<dyn Future>>`. `Pin<Box<dyn Future>>`  involves pinning and dynamic dispatch (which we need presumably since the future from `digest` can't be named. This is confusing stuff for relative newcomers!

Reply:
I tried expanding a little, about trying out a generic and realizing that wouldn't work, and a `Signing(impl Future)` isn't stable yet, and then finally asking Barbara for help again. 🤷